src/control/character_controller.rs

```diff
@@ -269,7 +271,13 @@ impl KinematicCharacterController {
                 character_shape,
                 ShapeCastOptions {
                     target_distance: offset,
-                    stop_at_penetration: false,
+                    stop_at_penetration:
+                    // We want to know the collision we're currently in,
+                    // otherwise we risk ignoring useful collision information (walls / ground...)
+                    iter_index == 0
+                    // If we stop at penetration while not allowing sliding alongside shapes, 
+                    // it will be impossible to get away of a collision.
+                    && self.slide,
```

Comment:
part 1 of the fix

Reply:
Actually, this fixes ground detection but only by side effect: the condition with `// No interference along the path.` contains a break, and `grounded` is computed just after. I think we actually want to computed the grounded status after the loop.